### ISSUE TITLE
Sprint 7

### ISSUE THREAD

**Issue by lastmjs** [lastmjs/podcrypt#209]
- [x] Downloading, playing, and adding to playlist should probably all be present...make the ... menu bigger
- [x] Make sure that the episode options work everywhere, sometimes when adding an episode not enough information is passed and everything gets messed up
- [x] Add the notification buttons back, so pausing and stuff works again
- [x] Use estimateGas to estimate gas for the transaction...test it on the intotheether address. Since it is a contract, it is requiring more than the 21000 gas required for simple funds transfers
- [x] The player needs to be redesigned again...it needs to be crisp and perfect. Switching between episodes is still not perfect...consider incorporating everything into Redux...there are so many asynchronous processes going on that it's hard to coordinate everything
- [x] Joe Rogan Experience #1309 throws a null error when downloading
- [x] Sometimes when ending an episode, the episode starts over and is not marked as listened for some reasons (this happens relatively often, still happening)
- [x] Add the ability to mark an episode not listened to
- [x] Downloading Hardcore History episodes is failing with a null error on Android and iOS...it seems to fetch the file just fine, but the IndexedDB storage fails
- [x] Watch this bug to see if we can salvage the audio element for playing downloaded files on iOS: https://bugs.webkit.org/show_bug.cgi?id=198015
- [x] Consider switching to the web audio API...it seems we will have even more flexibility if we need it down the road, and we might as we start adding advanced features later on. First determine if we can play audio from a blob using the web audio API. Then, make sure that we can still have the native media notifications on iOS and Android. If all goes well, consider making a custom element for this functionality...consider making it amazing and open sourcing it stand-alone
- [x] Study the web audio API deeply...I believe it may provide solutions to downloading and playing audio on iOS, and perhaps will help with some of the media notification issues on Chrome
- [x] URL.revokeObjectURL() to garbage collect object urls...it makes sense why the app was crashing so much when switching between downloaded episodes
- [x] Remove the set payout in progress action dispatch in the pc-wallet constructing thing
- [x] Podcast downloads
  - [x] The download ui does not work for some episode rows
  - [x] If you download an episode and immediately try to play it, it will break. I believe it still tries to use the episode.src directly because it has already been set
  - [x] Make sure downloadState is set correctly everywhere, especially when adding an episode to the playlist, right now it will be overriden
  - [x] Add the download option to all episode rows
  - [x] Show download indicator showing download progress
  - [x] Have an indicator if an episode is downloaded or not
  - [x] Allow the user to delete an episode
- [x] Add cursor pointer everywhere needed on desktop
- [x] The playlist episode url thing broke (I think I'm going to just get rid of that feature. Perhaps I can add it again in the future. A link to the episode description should be sufficient)
- [x] Update podcast ethereum address and ens name correctly in the wallet and everywhere...make sure it stays up to date correctly...for example, in the wallet, the address to display is pulled from the statically saved podcast description
- [x] fix podcrypt.eth resolver...the resolver contract address that I deployed is incorrect for some reason...I need to set it to the correct one (look at lastmjs.eth, that one was done correctly). After I setup the resolver, I can set the address for the resolver to resolve to
- [x] Fix podcrypt.eth resolver...it got messed up and set to 0x0x somehow...
- [x] Show the ENSName and the ethereumAddress to the user when the click on the verified button
- [x] Add copy link address options for podcasts and episodes
- [x] How I built this is broken
- [x] What bitcoin did is broken
- [x] Into the Ether is broken...their contract is taking up too much gas...should we increase the gaslimit? I guess so...
- [x] Add About button to the main page when the list of podcasts is empty. Make sure the about page is as simple and explanatory as possible
- [x] Open a bunch of bugs with Chromium for the notifications and potentially for CORS issues with mp3 files...why is cors enabled for mobile but not for desktop?
- [x] The z-indeces for the episode rows in the playlist and the player are off, the episode row is above the player in the playlist
- [x] Should all rows be the same size?
- [x] Loading the entire blob to play is causing memory issues, the Brave tab kept crashing when I was switching between episodes
- [x] When switching between episodes that had been downloaded, the time seemed to stay at what the previous episode was at...I'm pretty sure this was because of the enormous amount of memory the blob was taking up...so, we should load only parts of the arraybuffer that has been stored in the db...but, then we'll have to in-the-background switch out the source. iOS will most likely give us issues with this
- [x] Clean up numbers and strings, get rid of bignumber in as many places as possible, only use it for eth calculations
- [x] Type all actions
- [x] Time tracking is good but not perfect. When switching between episodes, the times start to get off by a bit
- [x] Switch to storing blobs for iOS if this bug pans out: https://bugs.webkit.org/show_bug.cgi?id=198278
  - [x] It looks like the issue has been resolved in at earliest iOS 12.2...decide if it's worth breaking all previous versions of iOS. I think it might be, since this app is being developed using the latest and greatest. By 1.0 hopefully most people will have a compatible device. Not sure though
  - [x] If we don't decide to switch to blobs, then we need to deal with the mimetype when creating the blob manually: We need to store the mime type of audio files in IndexedDB because of iOS. When creating the blob, we need to add the correct mime type. I am hard-coding the mime type to audio/mpeg when creating blobs right now, which should work for mp3 files but perhaps not for ogg or other audio formats.
- [x] Make sure the +10 -10 buttons are working, it seemed to perhaps not be...I skipped forward and thought I heard the same thing
- [x] Fix the slider being too far to the right
- [x] I get 404s for some episodes when trying to download
- [x] Make transition between downloaded episodes smooth...the loadeddata play seems to perhaps be messing with things when switching between downloaded episodes
- [x] Payments to Podcrypt duplicate many times over
- [x] When processing payments, what happens if an episode's rss feed can never be retrieved? I think then that the payment will fail over and over again
- [x] The downloading loading ui stopped working, probably because of the loading code I disabled
- [x] Times for episodes that don't have payouts during an interval do not get reset at the end of the interval I believe
- [x] Figure out chunking episodes...I believe that is the way to do it. This will get around a few issues that we've been having...chunk downloads, chunk storage, chunk playing. This should allow us to have many simultaneous downloads with files of any size (as long as total storage does not exceed the domain limit), and we should be able to load and play any file of any size that was stored.
- [x] Make sure that the app remembers the last route every single time it loads...sometimes it will reset you to the homepage after a restart, perhaps every time actually
- [x] Seeking forward and backward does not work when the player is paused...it does work, but the time does not update immediately
- [x] Make sure that the transition from streaming to downloading is smooth and works...if you are streaming an episode and then click download, it should transition perfectly
- [x] Render the episode overviews as html
- [x] Add a backup of all podcasts data...I'm getting nervous
- [x] Allow disabling podcasts...Think of the best way to do it, I want to still show the user if possible how much money the podcast is missing out on...not sure how to do that
- [x] Help out Speaking of Crypto, I think the feed is slightly messed up
- [x] Fix episode overviews, set the episode description from the content...
- [x] Figure out why episodes stored in the state have an undefined description
- [x] Make Podcrypt look good on Desktop...I'm in the middle of refactoring. I've just added in the test for desktop versus mobile screens, and now I'm changing the appropriate css in relation to that global property in the store. Also, while doing this refactor I should make sure that the rest of the app is responsive on all screen sizes in the Chrome thingy, and I should test it on real iOS devices. On some iOS devices Podcrypt isn't looking too sharp
  - [x] Fix podcast rows and episode rows (look good, uniform styling, cursor pointer)
  - [x] Fix the menu on desktop
  - [x] Fix the height thing on the main containers
- [x] I do not think that the playback rate is actually working
- [x] Rearrange icons on rows, make the 3 dots menu huge and on the left
- [x] Skipping forward or backward while paused is updating the times for the podcast...it should not do that
- [x] To get the duration and to be able to seek anywhere, we should load each chunk one at a time, decide its frame headers, calculate the duration of each frame based on length and nitrate, combine that to get the total duration of the chunk, then save the start and end time for that chunk with the chunk index ...do this for all chunks, and then we should have able to map times to chunks to load them on demand. We'll also have the total duration and should be able to calculate the current time... it's all coming together
- [x] Rename the download index stuff to chunk index
- [x] Way too many NaNs on the wallet
- [x] iOS Safari is blocking alerts...looks like we need modals/notification/confirmation much sooner
- [x] iOS backtracking before 0 the time gets all messed up
- [x] The progress event off of the audio element seems to be really slowing down scrolling on iOS
- [x] The search bar is not centered on iOS
- [x] The number inputs on iOS do not expand much, the width is too small. For example, 1000 shows as 10
- [x] iOS PWA bug...when adding to homescreen, localstorage is not copied over
- [x] The audio switching in the background between elements seems to potentially have problems on iOS...one time it died, and the audio context in the lock screen went away and the audio stopped playing. One time, with the screen on but locked, the audio kept repeating, as if it kept going back to the same spot over and over when one audio element ended
  - [x] It seems to keep repeating
  - [x] I think I should make a test page where I can test everything, all chunking and going back and forth between episodes. Then I can take that and test it quickly on many different devices
- [x] So it looks like Media Source Extensions is the way to go...instead of implementing the double audio element switching thing that I've got going on, switch to using Media Source Extensions. Hopefully it will handle memory better than URL.createObjectURL was directly with blobs. The problem is, it is not yet supported on iOS...so, hopefully URL.createObjectURL works better on iOS than it does on Android. I could just do that as a fallback if media extensions aren't present in a browser. This should solve the problem of current progress and total duration, but I'm not quite sure if the problem of switching tracks is solved yet. I might be able to add and remove buffers to do it though
- [x] The media source extensions seem to be working...I don't think the browser is crashing. But, I need to work with loading and such before playing, the loading does not seem to be working very well
  - [x] Seeking and sliding seems broken
  - [x] Sliding is coming along, just battling with how to modify the source buffer attached to the mediasource in real-time...I'm getting errors every way I try
- [x] Do not forget to do a bunch of migration work
- [x] Block everything that has to do with mediasource if on an ios device...or just check for mediasource itself, graceful upgrade then nothing has to really change once it is supported
- [x] As part of moving from beta to 1.0, I believe we need to refactor and make very solid the pc-player...perhaps use a local redux store...that's probably best actually, there is so much going on that I feel it is an application in and of itself, complex locking and such
- [x] Skipping forward and backward seems to work very well...sliding seems to have some locking issues still...sometimes it gets stuck when going back, but I think it's okay for beta
- [x] We might need to be revoking the object urls...I just had chrome crash after switching between episodes multiple times
- [x] mp4 files seem to be giving mse issues...mess around and see what you can do, try to get durations and playback for mp4 segments by blobbing, audio element, etc...if impossible, perhaps we can fallback to a full objectURL for any unsupported MSE filetypes and browsers. It won't be perfect, but we can do our best. Also, test manually MSE with lots of mp3 files. Let's find out if MSE is even suitable for for file types that it does support. I should also go gather the issues that I'm following on the media-source w3c repo, I've commented on or subscribed to some potentially relevant conversations. The all-around best solution still evades me. At all costs we want to avoid having to load the entire audio file into memory, since some of the files can be very large...but what if it's actually not an issue? Perhaps my phone is just kind of old, and in 1-2 years all phones will be able to support really large files being loaded into memory. Perhaps do some testing of objectURLs on other Android phones. If it doesn't crash them, then maybe it's alright? But then we still have the issue of switching tracks while in the background, I'm not sure how that would play out. I don't believe the double audio element thing is a good idea, because the transition was noticeable sometimes audibly, and it even caused the media notification to disappear and then reappear, not good. MSE still seems most elegant once iOS is supported and if mp4 files will ever be supported, both of which I'm not sure on. iOS should be supported soonish, ipadOS 13 will support MSE, so you'd think iOS would follow. But it's already been years, so I'm not sure I can count on it
- [x] New plan I'm considering, push off downloads until after 1.0...I can keep tabs on the relevant technologies and then integrate them once everything else is in place. So, downloads is not an MVP feature right now. I think this is correct. I've sunk a lot of time into trying to figure out downloads, and I've hit a wall in the technology. I can make a super polished experience everywhere else, with tests and integrations of DAI and have a solid foundation, hopefully with users and revenue. Then, when the technology is there, I can add downloads again. Shame really, I like listening to downloaded content...perhaps I'll just have to get used to buying more data
- [x] Generate the wallet immediately. In the about page, give podcast owners their Ethereum address to put into their description...perfect easy onboarding. We will then have to explain how to cash out through Coinbase or somewhere else. Maybe Wyre will have some kind of easy cash out solution eventually. Or hopefully Stripe or Fidelity or PayPal will just add Ethereum somehow
- [x] Let's talk ETC #88 just doesn't seem to be working for some reason
- [x] The new way to download seems to be working well on desktop and Android, though some episodes do not download properly for some reason...they are probably returning some kind of 404 or something that I am not handling properly
- [x] Chunk downloads again, make sure to serve range requests properly from the service worker fetch handler. We'll need to potentially grab up to two chunks from IndexedDB to serve up a range request correctly
- [x] When deleting a podcast, remove all downloads as well
- [x] Consider a streaming download instead of partial content
- [x] Backing up the state to a json file stopped partway for me...
- [x] Podnews cannot be downloaded
- [x] Make sure the confirmation prompt still shows up on iOS devices
- [x] Peacefully transition from a streaming episode to a downloaded episode
